src/Components/Components/src/ComponentFactory.cs

```diff
@@ -14,46 +15,76 @@ internal sealed class ComponentFactory
     private const BindingFlags _injectablePropertyBindingFlags
         = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
 
-    private static readonly ConcurrentDictionary<Type, Action<IServiceProvider, IComponent>> _cachedInitializers = new();
+    private static readonly ConcurrentDictionary<Type, (Action<IServiceProvider, IComponent> propertyInitializers, ObjectFactory? componentInitializer)> _cachedInitializers = new();
+    private readonly IComponentActivator? _componentActivator;
 
-    private readonly IComponentActivator _componentActivator;
-
-    public ComponentFactory(IComponentActivator componentActivator)
+    public ComponentFactory(IComponentActivator? componentActivator)
     {
-        _componentActivator = componentActivator ?? throw new ArgumentNullException(nameof(componentActivator));
+        _componentActivator = componentActivator;
     }
 
     public static void ClearCache() => _cachedInitializers.Clear();
 
     public IComponent InstantiateComponent(IServiceProvider serviceProvider, [DynamicallyAccessedMembers(Component)] Type componentType)
     {
-        var component = _componentActivator.CreateInstance(componentType);
-        if (component is null)
+        if (_componentActivator is not null)
         {
-            // The default activator will never do this, but an externally-supplied one might
-            throw new InvalidOperationException($"The component activator returned a null value for a component of type {componentType.FullName}.");
+            return InstantiateWithActivator(_componentActivator, serviceProvider, componentType);
         }
 
-        PerformPropertyInjection(serviceProvider, component);
+        return InstantiateDefault(serviceProvider, componentType);
+    }
+
+    private static IComponent InstantiateDefault(IServiceProvider serviceProvider, [DynamicallyAccessedMembers(Component)] Type componentType)
```

Comment:
Does the `DynamicallyAccessedMembers` make any difference here? The linker won't statically know anything about the `Type` parameter being passed, will it? I can see that on the public API (line 28) it might make a difference.

Perhaps you're putting it here just as a matter of correctness even if in practice it doesn't change the behavior.